src/vos/vos_tree.c

```diff
@@ -786,6 +739,98 @@ static struct vos_btr_attr vos_btr_attrs[] = {
 	},
 };
 
+static int
+tree_open_create(struct vos_object *obj, enum vos_tree_class tclass, int flags,
+		 struct vos_krec_df *krec, daos_handle_t *sub_toh)
+{
+	struct umem_attr	*uma = vos_obj2uma(obj);
+	struct vea_space_info	*info = obj->obj_cont->vc_pool->vp_vea_info;
+	daos_handle_t		 coh = vos_cont2hdl(obj->obj_cont);
+	int			 expected_flag;
+	int			 unexpected_flag;
+	int			 rc = 0;
+
+	if (flags & SUBTR_EVT) {
+		expected_flag = KREC_BF_EVT;
+		unexpected_flag = KREC_BF_BTR;
+	} else {
+		expected_flag = KREC_BF_BTR;
+		unexpected_flag = KREC_BF_EVT;
+	}
+
+	if (krec->kr_bmap & unexpected_flag) {
+		if (flags & SUBTR_CREATE) {
+			D_ERROR("Mixing single value and array not allowed\n");
```

Comment:
What's the reason behind this move?

Reply:
This is really moving code from rec_alloc to key_tree_prepare.  In rec_alloc, we don't know yet what the "type" is and I didn't really want to pass more information in the kbund/rbund.   So essentially, rather than creating the tree on insertion, I moved it to just after the insertion.  Note this also saves an open because the prior code did a create -> close -> open on first access whereas this either does open or create.

Reply:
I guess there was previously open code in key_tree_prepare but when I started adding create branches, it really wanted its own function.

Reply:
I thought a bit more about this last commet and it won't work with logging....because the tree persists across punches.   But as I said, Johann was fine with the restriction.